analyses/hello-clusters/01_perform-evaluate-clustering.Rmd

```diff
@@ -49,32 +62,443 @@ data_dir <- file.path(repository_base, "data", "current")
+cluster_results_df
+```
+
+
+```{r run on seurat}
+# Calculate clusters from a Seurat object using default parameters
+cluster_results_df <- calculate_clusters(seurat_obj)
+cluster_results_df
+```
+
+
+`rOpenScPCA` assumes that the PCA matrix is named `PCA` in SCE objects, and `pca` in Seurat objects (as they were in the above examples).
+If the PCA matrix you want to use in the object has a different name, you can provide the argument `pc_name`.
+We can see this below with an SCE object, for example:
+
+```{r use pc_name}
+# First, we'll rename the PCA matrix for demonstration
+reducedDimNames(sce) <- c("PCA_matrix", "UMAP")
+reducedDimNames(sce)
+
+# Calculate clusters from an SCE object using default parameters
+cluster_results_df <- calculate_clusters(
+  sce,
+  pc_name = "PCA_matrix"
+)
+cluster_results_df
+```
+
+
+
+## Saving clustering results 
+
+Results can either be directly exported as a TSV file (e.g., with `readr::write_tsv()`), or you can add the results into your SCE or Seurat object.
+The subsequent examples will show you how to save the `cluster_results_df` data frame to an SCE and a Seurat object.
+
+_A word of caution!_
+Objects from the ScPCA Portal already contain a column called `cluster` with results from an automated clustering.
+These automatic clusters were not evaluated, and their parameters were not optimized for any given library.
+
+To avoid ambiguity between the existing and new clustering results, you may want to rename incoming columns.
+
+We'll therefore begin by adding the prefix `ropenscpca_` to all column names in the `cluster_results_df` data frame before adding them to a given object.
+
+```{r rename columns}
+# First, rename columns in `cluster_results_df` to avoid ambiguity
+cluster_results_renamed_df <- cluster_results_df |>
+  # add the prefix openscpca_ to all columns
+  dplyr::rename_with(~ paste0("ropenscpca_", .x))
+
+cluster_results_renamed_df
+```
+
+
+### Saving results to an SCE object 
+
+Follow these steps to save the clustering results to an SCE object's `colData`:
+
+```{r add to sce}
+# First, extract the colData into a separate data frame for manipulation
+sce_coldata <- colData(sce) |>
+  as.data.frame()
+
+# Second, add columns as desired using the barcodes and cell_id columns to join
+sce_coldata <- sce_coldata |>
+  dplyr::left_join(
+    cluster_results_renamed_df,
+    by = c("barcodes" = "ropenscpca_cell_id")
+  )
+
+# Third, add the modified colData back into the SCE
+# Use the row.names argument to ensure rownames (barcodes) are retained; don't forget this step!
+colData(sce) <- DataFrame(
+  sce_coldata,
+  row.names = sce_coldata$barcodes
+)
+
+colData(sce)
+```
+
+### Saving results to a Seurat object 
+
+Follow these steps save the clustering results to a Seurat object's cell metadata:
+
+```{r add to seurat}
+# Add the cluster result data frame to the Seurat object metadata
+seurat_obj <- AddMetaData(seurat_obj, cluster_results_renamed_df)
```

Comment:
Does Seurat assume that the data frame is in the same order? If so, we might want to caution about that. 

In the interests of simplicity, you might also want to show making that assumption above, in which case you can first check that it is true with `all(colnames(sce) == cluster_results_df$cell_id)` or something like that, then just add the clusers more simply with `scr$ropenscpca_clusters <- cluster_results_df$cluster`. I don't think we really need to show adding all of the cluster parameters?